design/service-apis-implementation.md

```diff
@@ -0,0 +1,253 @@
+# Service-APIs implementation design
+
+Status: Accepted
+
+## Abstract
+The Service-APIs are the evolution of the Kubernetes APIs that relate to `Services`, such as Ingress.
+This document outlines what parts of these APIs Contour will implement, and how it will do so.
+
+## Background
+The Service-APIs are a subproject of Kubernetes SIG-Network, and are an attempt to re-do the mechanics around `Services` and `Ingress`, and how they interact.
+They are aiming to cover the things currently implemented by Layer 4 load balancers that implement `Services` of type `LoadBalancer`, and ingress controllers (like Contour).
+
+The Service APIs target three personas:
+- Infrastructure provider: The infrastructure provider (infra) is responsible for the overall environment that the cluster(s) are operating in. Examples include: the cloud provider (AWS, Azure, GCP, ...), the PaaS provider in a company.
+- Cluster operator: The cluster operator (ops) is responsible for administration of entire clusters. They manage policies, network access, application permissions.
+- Application developer: The application developer (dev) is responsible for defining their application configuration (e.g. timeouts, request matching/filter) and Service composition (e.g. path routing to backends).
+
+The cluster operator and application developer are basically the same as Contour's Cluster Administrator and Application Developer personas, which will be important for this design.
+
+In terms of the APIs themsleves, the Service APIs have 3 primary API resources (taken from the service-apis docs site):
+
+- GatewayClass defines a set of gateways with a common configuration and behavior.
+- Gateway requests a point where traffic can be translated to Services within the cluster, using an internal Listener construct.
+- Routes describe how traffic coming via the Gateway maps to the Services.
+
+In Contour, we've previously solved a lot of the same problems with HTTPProxy (and IngressRoute before it).
+That functionality can be described in the Service APIs by HTTPRoutes and TLSRoutes, as they describe Layer 7 ingress, as Contour does today.
+
+Other types of routes include TCPRoutes and UDPRoutes, which are intended for Layer 4 load balancers. Implementations may also define their own Route objects, at the cost of interoperability.
+
+In terms of its stated goals, Contour is aiming at being an ingress controller - that is, a Layer 7 proxy with some api gateway functions.
+Currently, Contour provides "TCP Proxying" that allows the forwarding of TLS streams based on the SNI information, which is precisely what the Service APIs TLSRoute object is for.
+If Project Contour (the organisation) does add support for TCP and UDP forwarding, it will not be in the `projectcontour/contour` repo, but will be a separate repo.
+
+## Goals
```

Comment:
is it a goal for Contour's service-apis support to reach ~feature parity with `HTTPProxy`? If so, how will we implement features that aren't explicitly part of the service-apis spec? And if not, what set of features _do_ we expect to implement?

Another way of asking this is, are we going to use annotations to implement features that aren't directly a part of service-apis, like we do with Ingress?

Reply:
I'll put something in the background to summarise this, but:
- I think the goal should absolutely be that eventually service-apis has feature parity with HTTPProxy
- the upstream does not have support for many features right now
- we should implement what is implementable right now
- we should collaborate with upstream to work out what features are part of the core, extension, and implementation specific sections of service-apis.

The service-apis are specifically built with extension points so you don't have to use annotations. That's what the reference points that allow you to point to any object, but default to a specific kind, are for.